### ISSUE TITLE
负载均衡损失的两种实验结果

### ISSUE THREAD

**Issue by github-luffy** [jingyaogong/minimind#635]
大佬，有MoE的两种负载均衡损失的实验结果吗，哪个负载均衡损失会好点？个人觉得应该是Token级别的负载均衡损失好。
if self.seq_aux:
    scores_for_seq_aux = scores_for_aux.view(bsz, seq_len, -1)
    ce = torch.zeros(bsz, self.n_routed_experts, device=hidden_states.device)
    ce.scatter_add_(1, topk_idx_for_aux_loss,
                    torch.ones(bsz, seq_len * aux_topk, device=hidden_states.device)).div_(
        seq_len * aux_topk / self.n_routed_experts)
    aux_loss = (ce * scores_for_seq_aux.mean(dim=1)).sum(dim=1).mean() * self.alpha
else:
    mask_ce = F.one_hot(topk_idx_for_aux_loss.view(-1), num_classes=self.n_routed_experts)
    ce = mask_ce.float().mean(0)
    Pi = scores_for_aux.mean(0)
    fi = ce * self.n_routed_experts
    aux_loss = (Pi * fi).sum() * self.alpha

**Comment by jingyaogong** [comment]
很好的问题
不考虑shared expert的话，目前有 4 个专家
假设 1 个 batch 有 2 个样本，每个样本 3 个 token：
```
样本A: [token1, token2, token3] → 专家分配 [1, 1, 1] (全去专家1)
样本B: [token4, token5, token6] → 专家分配 [2, 3, 4] (均匀分布)
```

### Token 级别（seq_aux=False）

做法：把所有 6 个 token 混在一起统计
分布: 专家1=3次, 专家2=1次, 专家3=1次, 专家4=1次 → 计算全局 aux_loss

优点：简单一些
缺点：样本 A 全部偏向专家 1，但被样本 B 的均匀分布“稀释”了，aux_loss 看起来还行，但 A 实际上严重不均衡。

### Sequence 级别（seq_aux=True）

做法：每个样本独立计算，再平均
样本A: 专家1=3次, 其他=0次 → loss_A 很大（严重不均衡）
样本B: 均匀分布 → loss_B 很小
最终 aux_loss = (loss_A + loss_B) / 2

优点：样本 A 的不均衡问题不会被样本 B 掩盖，能更准确地惩罚局部不均衡。
缺点：复杂一些

不过，即使理论如此。实际效果2种差异微乎其微，在 Batch (也许>=8时) 训练就足够趋于一致了，甚至之后可以直接简化去掉seq_aux这个参数。

---

<img width="950" height="604" alt="Image" src="https://github.com/user-attachments/assets/15042123-5aed-4e73-a298-c27c4c63a141" />

话说回来，aux-loss在这种少专家的情况下，距离考虑以上问题其实中间还有很多已知问题，举例1:
负载本身
1、专家负载均衡与任务性能是有冲突的，aux-loss 强制专家均匀分配，但「最优路由」不可能绝对均匀，强制用大的均衡惩罚反而会破坏 router 学到的「有意义的专业化分工」，相当于给模型加了约束，限制了表达能力
2、辅助负载 loss 主要目的是防止专家坍缩（所有 token 都去少数专家），但是在如此少的4个专家数量和如此单薄的数据量时，我认为aux-loss 的「约束」大于「保护」作用

从这个规模的实验来看，MoE with Aux-loss 反而整体学出来logits-loss不如MoE w/o Aux-loss（不额外限制专家均衡，靠router的天然反传学路由）

<img width="1984" height="862" alt="Image" src="https://github.com/user-attachments/assets/e47ff35e-6256-459e-a6e5-675917fc1be7" />

另外我设置的router_aux_loss_coef=5e-4，topk=1，8层网络，理论值MoE aux-loss是收敛到4e-3，这一点是完美符合的（此时是token level，其实都一样可以忽略）

$$\mathrm{aux\_loss} = \mathrm{top\_k} \times \mathrm{coef} \times \mathrm{layers}$$
